File: spots_tracker.cpp

#include <algorithm>

#include "spots_tracker.hpp"

const double INFTY = 1e100;

Spot::Spot() {

}

Spot::Spot(Point2f center, double weight)
  : center(center), weight(weight) {

}

SpotNode::SpotNode(Spot spot, double time, int num, bool present)
  : spot(spot), time(time), num(num), present(present), badness(INFTY), prev(NULL) {

}

SpotsTracker::SpotsTracker(control_panel_t &panel)
  : panel(panel),
    node_num(0),
    front_num(-1),

    dynamic_depth(60),
    back_badness(INFTY),
    appearance_badness(400.0),
    disappearance_badness(400.0),
    absence_badness(-10.0),
    max_speed(18.0),
    max_unseen_distance(0.3),
    skip_badness(-8.0),
    variance_parameter(0.3) {

  // Push a frame with just an absent node; it will be ignored when
  // returning results
  SpotNode phantom_node(Spot({0.0, 0.0}, 0.0), 0.0, -1, false);
  phantom_node.badness = 0.0;
  vector< SpotNode > v;
  v.push_back(phantom_node);

  this->frame_num_rev_map[-1] = 0.0;
  this->timeline.push_back(v);

}

tuple< bool, double > SpotsTracker::jump_badness(const SpotNode &n1, const SpotNode &n2) const {

  double ret = 0.0;
  double time = n2.time - n1.time;
  double skip = n2.num - n1.num - 1;
  assert(time >= 0);
  assert(skip >= 0);

  // Presence transitions
  if (n1.present && !n2.present) {
    // Present to absent
    ret += this->disappearance_badness;
  } else if (!n1.present && n2.present) {
    // Absent to present
    ret += this->appearance_badness;
  } else if (!n1.present && !n2.present) {
    // Absent to absent
    ret += this->absence_badness;
    if (skip > 0) return make_tuple(false, 0.0);
  }

  // Node weight
  if (n2.present) {
    ret -= n2.spot.weight;
  }

  // Skip badness
  ret += skip * this->skip_badness;

  // Locality check and Gaussian likelihood
  if (n1.present && n2.present) {
    double distance = norm(n1.spot.center - n2.spot.center);
    if (distance > this->max_speed * time) return make_tuple(false, 0.0);
    if (distance > this->max_unseen_distance) return make_tuple(false, 0.0);
    ret += distance * distance / (time * this->variance_parameter);
  }

  return make_tuple(true, ret);

}

void SpotsTracker::push_back(vector< Spot > spots, double time) {

  // Remember the reverse map from frame_num to time
  this->frame_num_rev_map[this->node_num] = time;

  // When first frame is pushed, put its timestamp to the initial
  // phantom frame (so that time counte does not explode)
  if (this->timeline.front().front().num == -1) {
    this->timeline.front().front().time = time;
  }

  // Prepere a vector with the new nodes
  vector< SpotNode > v;
  for (auto spot : spots) {
    v.emplace_back(spot, time, this->node_num, true);
  }
  v.emplace_back(Spot({0.0, 0.0}, 0.0), time, this->node_num, false);
  this->node_num++;

  long int begin = std::max< long int >(0, this->timeline.size() - this->dynamic_depth);
  this->back_badness = INFTY;
  for (vector< SpotNode >::iterator it2 = v.begin(); it2 != v.end(); it2++) {
    SpotNode &n2 = *it2;

    // Implement a step of the dynamic programming algorithm (with
    // bounded depth)
    for (long int i = begin; i < this->timeline.size(); i++) {
      for (vector< SpotNode >::iterator it1 = this->timeline[i].begin(); it1 != this->timeline[i].end(); it1++) {
        SpotNode &n1 = *it1;
        bool legal_jump;
        double delta_badness;
        tie(legal_jump, delta_badness) = this->jump_badness(n1, n2);
        if (!legal_jump) continue;
        double new_badness = n1.badness + delta_badness;
        if (new_badness < n2.badness) {
          n2.badness = new_badness;
          n2.prev = it1;
          n2.prev_num = n1.num;
          n2.prev_time = n1.time;
          n2.prev_present = n1.present;
          n2.prev_spot = n1.spot;
        }
      }
    }

    // Update the best position in the back of the queue
    if (n2.badness < this->back_badness) {
      this->back_badness = n2.badness;
      this->back_best = it2;
    }
  }

  // Push the vector
  this->timeline.push_back(v);

}

static Point2f interpolate(double a, double t, double b, Point2f x, Point2f y) {

  double t1 = (t - a) / (b - a);
  return (1 - t1) * x + t1 * y;

}

SpotNode &SpotsTracker::_front() {

  // Go backward until you find the front level
  vector< SpotNode >::iterator it = this->back_best;
  while (it->prev_num > this->front_num) {
    it = it->prev;
  }
  return *it;

}

tuple< bool, Point2f > SpotsTracker::front() {

  // Return _front(), unless it has number -1; in that case try to pop
  if (this->front_num == -1) {
    this->_pop_front();
  }
  SpotNode &ret = this->_front();

  if (ret.prev_num == this->front_num) {
    return make_tuple(ret.prev_present, ret.prev_spot.center);
  } else if (ret.num == this->front_num) {
    return make_tuple(ret.present, ret.spot.center);
  } else {
    if (ret.present && ret.prev_present) {
      return make_tuple(true, interpolate(ret.prev_time, this->frame_num_rev_map[this->front_num], ret.time, ret.prev_spot.center, ret.spot.center));
    } else {
      return make_tuple(false, ret.spot.center);
    }
  }

}

void SpotsTracker::_pop_front() {

  this->timeline.pop_front();
  this->frame_num_rev_map.erase(this->front_num);
  this->front_num += 1;

  assert(this->front_num == this->timeline.front().front().num);

}

void SpotsTracker::pop_front() {

  if (this->front_num == -1) this->_pop_front();
  this->_pop_front();

}

int SpotsTracker::get_front_num() {

  return this->front_num;

}

File: spots_tracker.py
"""
Implementation of Gio-Giove algorithm to reconstruct ball trajectory.
"""

import collections
import itertools
import cv2



class Spot:
    """
    A relevant point of some frame.
    """
    
    def __init__(self, point, weight):
        self.point = point
        self.weight = weight
        self.frame_num = None  # Frame number (possibly set by Layer)
        self.time = None    # Layer time (set by Layer)
        
        self.badness = None # Badness from starting spot to this spot (will be set by SpotsTracker.best_trajectory)
        self.prev_spot = None   # Previous spot in dynamic programming reconstruction (will be set by SpotsTracker.best_trajectory)
    
    def present(self):
        return self.point is not None


class Layer:
    """
    Contains all relevant information about a frame.
    """
    
    def __init__(self, spots, frame_num, time):
        self.spots = spots  # List of spots
        
        # Add the special Spot with absent ball
        self.spots.append(Spot(None, SpotsTracker.ABSENT_BALL_WEIGHT))
        
        # Set frame_num and time for the spots
        for spot in spots:
            spot.frame_num = self.frame_num
            spot.time = self.time
        

class SpotsTracker:
    
    ABSENT_BALL_WEIGHT = 0.0    # Weight of a Spot with absent ball
    SKIP_BADNESS = 0.0  # Badness added for every skipped frame
    
    DISAPPEARANCE_BADNESS = 400.0
    APPEARANCE_BADNESS = 400.0
    ABSENCE_BADNESS = -10.0
    
    MAX_SPEED = 18.0
    MAX_UNSEEN_DISTANCE = 0.3
    
    VARIANCE_PARAMETER = 0.3
    
    
    def __init__(self):
        self.timeline = collections.deque() # Deque of layers
        self.first_num = None   # frame_num of the first layer in the timeline
        self.last_num = None    # frame_num of the last layer in the timeline
        
        self.fake_initial_spot = Spot(None, 0.0)    # Used in dynamic programming
        self.fake_final_spot = Spot(None, 0.0)       # Used in dynamic programming
    
    
    def push_back(self, layer):
        self.timeline.append(layer)
        
        # Update self.last_num
        self.last_num = layer.frame_num
        
        # Update self.first_num if the timeline was empty
        if len(self.timeline) == 1:
            self.first_num = self.timeline[0].frame_num
        
        # Update fake_final_spot
        self.fake_final_spot.frame_num += 1
        
        # Update fake_initial_spot.frame_num if it is None
        if self.fake_initial_spot.frame_num is None:
            self.fake_initial_spot.frame_num = self.timeline[0].frame_num - 1
        
        # Do dynamic programming for the new layer and for fake_final_spot
        for spot in itertools.chain(layer.spots, [self.fake_final_spot]):
            spot.badness = None
            spot.prev_spot = None
            
            # Find best trajectory until this spot, trying all possible previous spots
            for previous in self.extended_spots():
                if previous.frame_num < spot.frame_num:
                    # Try arc from previous to spot
                    badness = self.arc_badness(previous, spot)
                    if badness is not None:
                        badness += previous.badness
                        if spot.badness is None or badness < spot.badness:
                            # New badness is lower!
                            spot.badness = badness
                            spot.prev_spot = previous
        
    
    def pop_front(self):
        self.timeline.popleft()
        
        # Update self.first_num
        if len(self.timeline) == 0:
            self.first_num = None
        else:
            self.first_num = self.timeline[0].frame_num
    
    
    def extended_spots(self):
        """
        Returns a generator of all the spots in the timeline, including fake_initial_spot and fake_final_spot.
        """
        yield self.fake_initial_spot
        for layer in self.timeline:
            for spot in layer.spots:
                yield spot
        yield self.fake_final_spot
    
    
    def arc_badness(self, start, end):
        """
        Returns the badness of the arc from Spot start to Spot end, or None if the arc does not exist.
        """
        time = end.time - start.time
        skip = end.frame_num - start.frame_num - 1
        result = 0.0
        
        # Presence transitions
        if start.present() and not end.present():
            # Present to absent
            result += self.DISAPPEARANCE_BADNESS
        elif not start.present() and end.present():
            # Absent to present
            result += self.APPEARANCE_BADNESS
        elif not start.present() and not end.present():
            # Absent to absent
            result += self.ABSENCE_BADNESS
            if skip > 0:
                return None
        
        # Node weight
        result -= end.weight
        
        # Skip badness
        result += float(skip) * self.SKIP_BADNESS
        
        # Locality check and gaussian likelihood
        if start.present() and end.present():
            distance = cv2.norm(start.point - end.point)
            if distance > self.MAX_SPEED * time:
                return None
            if distance > self.MAX_UNSEEN_DISTANCE:
                return None
            result += distance ** 2 / (time * self.VARIANCE_PARAMETER)
        
        return result
    
    
    def interpolate(self, first_spot, second_spot, time):
        a = first_spot.time
        b = second_spot.time
        k = (time - a) / (b - a)
        return (1 - k) * first_spot.point + k * second_spot.point
    
    
    def estimate_position(self):
        """
        Returns the estimated position of the ball for the first layer in the timeline, or None if the ball appears to be out of the field.
        """
        spot = self.fake_final_spot
        previous = spot.prev_spot
        
        while previous is not None and previous.frame_num > self.first_num:
            spot = previous
            previous = spot.prev_spot
        
        if previous is None or previous is self.fake_initial_spot or spot is self.fake_final_spot:
            # The ball appears to be out of the field
            return None
        
        elif previous.frame_num == self.first_num:
            # The estimate is done for the correct layer
            return previous.point
        
        else:
            # Interpolate between the two closest estimates
            return self.interpolate(previous, spot, self.layers[0].time)
    
    
    def best_trajectory(self):
        """
        Returns a list of spots (at most one per layer) that is the best trajectory for the ball, and the badness of such trajectory.
        """
        
        for spot in self.extended_spots():
            if spot is self.fake_initial_spot:
                # Base step
                spot.badness = 0.0
                spot.prev_spot = None
            else:
                # Dynamic programming
                spot.badness = None
                spot.prev_spot = None
                
                # Find best trajectory until this spot, trying all possible previous spots
                for previous in self.extended_spots():
                    if previous.frame_num < spot.frame_num:
                        # Try arc from previous to spot
                        badness = self.arc_badness(previous, spot)
                        if badness is not None:
                            badness += previous.badness
                            if spot.badness is None or badness < spot.badness:
                                # New badness is lower!
                                spot.badness = badness
                                spot.prev_spot = previous
        
        backward_trajectory = []
        spot = self.fake_final_spot
        while spot.prev_spot is not None and spot.prev_spot is not self.fake_initial_spot:
            spot = spot.prev_spot
            backward_trajectory.append(spot)
        
        trajectory = reversed(backward_trajectory)
        badness = self.fake_final_spot.badness
        
        return trajectory, badness


